content/blog/2016/2016-08-29-sauce-pipeline.adoc

```diff
@@ -0,0 +1,271 @@
+---
+
+NOTE: This is a guest post by link:https://github.com/bitwiseman[Liam Newman],
+Technical Evangelist at Cloudbees.
+
+Testing web applications across multiple browsers on different platforms can be challenging even for smaller applications.
+With Jenkins and the
+link:https://wiki.jenkins-ci.org/display/JENKINS/Sauce+OnDemand+Plugin[Sauce OnDemand Plugin],
+you can wrangle that complexity by defining your Pipeline as Code.
+
+
+=== Pipeline &#9829; UI Testing, Too
+I recently started looking for a way to do browser UI testing for an open-source JavaScript project.
+The project is targeted primarily at
+link:https://nodejs.org[Node.js]
+but we're committed to maintaining browser-client compatibility as well.
+That means we should run tests on a matrix of browsers.
+link:https://saucelabs.com/[Sauce Labs]
+has an "open-sauce" program that provides free test instances to open-source projects.
+I decided to try using
+link:https://wiki.jenkins-ci.org/display/JENKINS/Sauce+OnDemand+Plugin[Sauce OnDemand Plugin]
+and
+link:http://nightwatchjs.org/[Nightwatch.js]
+to run Selenium tests on sample project first, before trying a full-blown suite of tests.
+
+=== Starting from Framework
+I started off by following Sauce Labs' instructions on
+"link:https://wiki.saucelabs.com/display/DOCS/Setting+Up+Sauce+Labs+with+Jenkins[Setting up Sauce Labs with Jenkins]"
+as far as I could.
+I installed the
+link:https://wiki.jenkins-ci.org/display/JENKINS/JUnit+Plugin[JUnit] and
+link:https://wiki.jenkins-ci.org/display/JENKINS/Sauce+OnDemand+Plugin[Sauce OnDemand]
+plugins, created an account with Sauce Labs, and added my Sauce Labs credentials to Jenkins.
+From there I started to get a little lost.
+I'm new to Selenium and I had trouble understanding how to translate the instructions to my situation.
+I needed a working example that I could play with.
+
+Happily, there's a whole range of samples in the github org
+link:https://github.com/saucelabs-sample-test-frameworks[saucelabs-sample-test-frameworks],
+which show how to integrate Sauce Labs with various test frameworks, including Nightwatch.js.
+I forked the Nightwatch.js sample to
+link:https://github.com/bitwiseman/JS-Nightwatch.js[bitwiseman/JS-Nightwatch.js]
+and set to writing my `Jenkinkfile`.
+Between the sample and the Sauce Labs instructions,
+I was able to write a pipeline that ran five tests on one browser via Sauce Connect:
+
+[source,groovy]
+----
+node {
+    stage "Build"
+    checkout scm
+
+    sh 'npm install' // <1>
+
+    stage "Test"
+    sauce('f0a6b8ad-ce30-4cba-bf9a-95afbc470a8a') { // <2>
+        sauceconnect(options: '', useGeneratedTunnelIdentifier: false, verboseLogging: false) { // <3>
+            sh './node_modules/.bin/nightwatch -e chrome --test tests/guineaPig.js || true' // <4>
+            junit 'reports/**' // <5>
+            step([$class: 'SauceOnDemandTestPublisher']) // <6>
+        }
+    }
+}
+----
+<1> Install dependencies
+<2> Use my sauce credentials
```

Comment:
I think this should be more verbose. This is the ID of the Credentials stored in Jenkins right? Perhaps there is (lord I hope so) a documentation page somewhere to link to for "Adding Credentials?"